### DISCUSSION TITLE
System reboot/restart handling coverage

### DISCUSSION THREAD

**Discussion by sykim-etri** [cloud-barista/cb-ladybug#40] [Ideas]
#39 와 관련하여 CB-Ladybug에서 사용 중인 시스템 자원들의 reboot 또는 restart에 대해 어떻게 대응할지에 대해 논의합니다.

**Comment by sykim-etri** [comment]
용어 정의 관련 @powerkimhub 의견
* reboot(재부팅): VM이 바로 start 될 것이므로, publicip를 해제하지 않음. 이슈 없음.
* restart(stop->start): 사용자가 stop 후 원하는 시점에 다시 start,  언제 VM이 start될지 모르므로 publicip 회수.
  * known-restart: Cloud-Barista API를 이용한 restart
  * unknown-resart: CSP Console 또는 SSH 사용하여 VM을 직접 restart
저는 이런 의미로 사용했었는데, 오해의 소지가 있을 수 있겠습니다.

**Reply by sykim-etri** [reply]
@seokho-son 의견

* reboot(재부팅): VM이 바로 start 될 것이므로, publicip를 해제하지 않음. 이슈 없음. ==> 클라우드별로 상황이 다르므로, 이렇게 
단정하기 어려울 것 같습니다.
* restart(stop->start): 사용자가 stop 후 원하는 시점에 다시 start,  언제 VM이 start될지 모르므로 publicip 회수.
  * known-restart: Cloud-Barista API를 이용한 restart
  * unknown-restart: CSP Console 또는 SSH 사용하여 VM을 직접 restart ==> unknown-reboot의 상황도 있으므로 구분하는 것이 좋을 것 같습니다. 특히, 사용자가 SSH를 통해 VM을 stop 했다가 SSH를 통해 start하는 것은 불가능합니다.

생각보다 다양한 상황이 있으므로, 다시 정리해보자면 아래와 같습니다.

* OS-reboot: VM OS 수준의 재부팅 (Public IP와 관계 없음)}
* TB-Reboot: TB를 통한 재부팅 (클라우드 타입에 따라 Public IP가 일부 변경. OS-reboot이 되는 경우도 있고, Suspend->Resume 처리가 되기도 함)
* TB-Restart: TB를 통해 TB-Suspended VM을 Resume 재구동 (Public IP 변경)
* CSP-Reboot: CSP 콘솔에서 사용자가 직접 VM을 Reboot (Public IP 변경 가능성 있음)
* CSP-Restart: CSP 콘솔에서 사용자가 직접 Suspended된 VM을 Resume (Public IP 변경)
(참고로, Restart 는 TB에 정의된 MCIS 라이프사이클이 아닙니다. TB에서는 이를 Suspend와 Resume으로 분리되어 있습니다.)
  * known-reboot: TB-Reboot
  * known-restart: TB-Restart
  * unknown-reboot: CSP-Reboot, OS-reboot
  * unknown-restart: CSP-Restart, OS-restart (불가능한 제어) (edited)

**Reply by sykim-etri** [reply]
@powerkimhub 의견

publicip와 관련된 이벤트 종류는 아래처럼 분류 및 정의하면 될 것 같습니다.
* TB-Reboot
* TB-Restart
* CSP-Reboot/OS-reboot
* CSP-Restart

<의견> publicip 혹은 연장선에서
* MCIS reboot/restart 연관된 궁금한 점
  * MCIS reboot/restart를 정상 operation으로 인지할 것인지?
    * 현재 MCKS 생성 모델에서는 MCKS 생성 요청시 내부에서 자동으로 생기는 시스템 MCIS입니다.
    * 이들 MCIS들을 사용자가 일반 MCIS 수준으로  제어할수 있게 보여줄것인지? 아니면, 관리 목적으로만 허용할 것인지?
  * MCIS가 start되면, 자동으로 MCKS를 가동시킬 것인지? MCKS를 만든 사용자가 선택적으로 가동시킬것인지?
    * 이 선택은 위의 '사용자에게 보여줄것인지' 선택에 의존적일 수 있겠습니다.
* publicip 등 restart시 VM 변경 정보관련
  * MCKS 최초 생성시 PublicIP: 이슈 없음.
  * MCKS 재가동시 PublicIP: 변경되면 기존 배포된 MCKS 동작에 문제 있음.
  * MCKS 최초 생성시 K8S SW 설치와 설정을 최대한 분리하면 좋을 듯합니다.
    * 예시) 1단계: MCIS 생성, 2단계: MCKS SW 배포, 3단계: MCKS 설정(publicip 등), 4단계: MCKS 마스터 가동, 5단계: MCKS Worker 가동/등록, 6단계: MCKS 상태 체크
    * 이렇게 분리하고 나면,
      * 최초 생성시: 1~6 단계 수행
      * 재가동시(MCKS SW 이미 설치): 3~6단계 수행
      * 1-2단계가 cost가 가장 비싸고 오래걸릴거라 생각합니다. 어떤 식으로든 1~2단계는 재활용되어야 할 거 같습니다.
    * 이럴려면,
      * 기존 pkg 관리 플랫폼(apt 등)을 사용하지 않아야 할수 있겠습니다.
      * SW 설치와 publicip 등의 setup을 뜯어내야 하겠습니다.
      * 처음에 시간이 걸리겠지만,
      * 이렇게 단계 구분을 해둬야 나중에 MCAS 등에서 MCKS pool 등을 ladybug 내부에서 운영할수 있을 거 같습니다.
    * 이렇게 분리한다고 하면,
      * publicIP를 ladybug에서 고민할 필요가 없을 듯 합니다.

<의견> API관련
* LB는 MCKS뿐만 아니라, MCAS에 대한 서비스를 제공합니다.
* 현재 API path는 /~/clusters로 정의되어 있습니다. 아래와 같은 고려가 필요해 보입니다.
  * /~/mcks/{mcks}/clusters
  * /~/mcas/{mc-app}
* 이미 고려가 되어 협의 및 정리가 된 사항이라면 유지하셔도 괜찮겠습니다.
* 아직은 API 활용이 확산된 상황은 아니니, 시간을 가지고 고민해보시기 바랍니다.

**Reply by sykim-etri** [reply]
> @powerkimhub 의견
>   * MCIS reboot/restart를 정상 operation으로 인지할 것인지?
>     
>     * 현재 MCKS 생성 모델에서는 MCKS 생성 요청시 내부에서 자동으로 생기는 시스템 MCIS입니다.
>     * 이들 MCIS들을 사용자가 일반 MCIS 수준으로  제어할수 있게 보여줄것인지? 아니면, 관리 목적으로만 허용할 것인지?
>   * MCIS가 start되면, 자동으로 MCKS를 가동시킬 것인지? MCKS를 만든 사용자가 선택적으로 가동시킬것인지?
>     
>     * 이 선택은 위의 '사용자에게 보여줄것인지' 선택에 의존적일 수 있겠습니다.
> * publicip 등 restart시 VM 변경 정보관련

CB-TB에서는 MCKS의 존재를 알수 없기 때문에 사용자에게 일반MCIS와 MCKS용 MCIS를 구분하여 보여줄 방법이 없을 것으로 생각됩니다. 따라서 TB-Reboot/Restart는 정상적인 동작으로 인지하고 대응하는 것이 바람직할 것으로 보입니다.
@itnpeople 관련하여 의견 부탁드립니다.

**Comment by seokho-son** [comment]
저는.. systemd 를 활용하여 클러스터를 재구성하는 현재의 방식도
systemd 가 가능한 os 로 MCKS의 host를 한정할 수 있다면,

빠른 클러스터 복구 관점에서, 적절한 대응 방법이라고 생각합니다.
(Ladybug에서 선제적으로 매우 잘 진행해주신 사항이라고 봅니다.)

단, 해당 사항은 사용자에 의한 OS 수준의 reboot(OS-reboot)이 아니라, (Public IP가 변경되지 않으므로)

TB 또는 CSP콘솔을 통해 Suspend 및 Resume (또는 Reboot)하는 경우(TB-Reboot, TB-Restart / CSP-Restart)에 활용될 것 같습니다.
(TB-Reboot, TB-Restart / CSP-Restart 액션은 VM의 Public IP를 변경시킬 수 있음)

일단, 현재 PR은 수용하고, 향후 해당 기능을 config를 통해 enable / disable 할 수 있도록 개선해 두는 것은 어떨까요..?

**Reply by itnpeople** [reply]
원래 제 의도는 라이프사이클 관리 측면을 생각 했다기 보다는 
VM이 유동IP 체제이니 최소한의 신뢰성 보장 해보자는 의도였다고 말씀드리고 싶네요.
개발측면에서 말씀드리자면
network-cni 별로 ip alias + kubelet + annotation 작업이 공통적으로 들어가야 클러스터가 구성될 수 있으므로 
공통화하고 가능하면 ip가 바뀐는 시점에도 대응해보자...
였습니다.
